sdk/common/src/main/scala/org/typelevel/otel4s/sdk/Resource.scala

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2023 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.typelevel.otel4s.sdk
+
+import cats.Show
+import cats.implicits.catsSyntaxEitherId
+import cats.implicits.catsSyntaxOptionId
+import cats.implicits.catsSyntaxSemigroup
+import cats.implicits.showInterpolator
+import org.typelevel.otel4s.Attribute
+import org.typelevel.otel4s.sdk.BuildInfo
+import org.typelevel.otel4s.sdk.Resource.ResourceInitiationError
+import org.typelevel.otel4s.semconv.resource.attributes.ResourceAttributes._
+
+import scala.util.control.NoStackTrace
+
+/** [[Resource]] serves as a representation of a resource that captures
+  * essential identifying information regarding the entities associated with
+  * reported signals, such as statistics or traces.
+  * @param attributes
+  *   \- a collection of [[Attribute]]s
+  * @param schemaUrl
+  *   \- an optional schema URL
+  */
+final case class Resource(
+    attributes: Attributes,
+    schemaUrl: Option[String]
+) {
+
+  /** Merges [[Resource]] into another [[Resource]]. If the same attribute
+    * exists in both resources, the attribute in the other [[Resource]] will be
+    * used.
+    * @param other
+    *   \- the other [[Resource]] to merge into.
+    * @return
+    *   a new [[Resource]] with the merged attributes.
+    */
+  def mergeInto(other: Resource): Either[ResourceInitiationError, Resource] = {
+    if (other == Resource.Empty) this.asRight
+    else {
+      val schemaUrlOptEither = (other.schemaUrl, schemaUrl) match {
+        case (Some(otherUrl), Some(url)) =>
+          if (otherUrl == url)
+            url.some.asRight
+          else
+            ResourceInitiationError.SchemaUrlConflict.asLeft
+        case (otherUrl, url) =>
+          otherUrl.orElse(url).asRight
+      }
+
+      schemaUrlOptEither.map(
+        Resource(
+          other.attributes |+| attributes,
+          _
+        )
+      )
+    }
+  }
+
+  /** Unsafe version of [[Resource.mergeInto]] which trows an exception if the
+    * merge fails.
+    */
+  def mergeIntoUnsafe(other: Resource): Resource =
+    mergeInto(other).fold(
+      throw _,
+      identity
+    )
+}
+
+object Resource {
+  sealed trait ResourceInitiationError extends NoStackTrace
```

Comment:
Are we sure that a stack trace isn't helpful in this situation?
```suggestion
  sealed abstract class ResourceInitiationError extends NoStackTrace
```